com/williamfiset/datastructures/binarysearchtree/SplayTree.java

```diff
@@ -0,0 +1,274 @@
+
+            }
+            return (root.getLeft() == null) ? root : rightRotate(root);
+        } else {
+            if (root.getRight() == null) return root;
+
+            if (root.getRight().getData().compareTo(key) > 0) {
+                root.getRight().setLeft(splayUtil(root.getRight().getLeft(), key));
+        if (root.getRight().getLeft() != null)
+                    root.setRight(rightRotate(root.getRight()));
+            } else if (root.getRight().getData().compareTo(key) < 0)// Zag-Zag (Right Right)
+            {
+                root.getRight().setRight(splayUtil(root.getRight().getRight(), key));
+                root = leftRotate(root);
+            }
+
+            return (root.getRight() == null) ? root : leftRotate(root);
+        }
+
+    }
+
+    public BinaryTree<T> splay(T x) {
+        if (root == null) return null;
+
+        this.root = splayUtil(root, x);
+
+        return this.root;
+    }
+    public BinaryTree<T> search(T x) {
+        if (root == null) return null;
+
+        this.root = splayUtil(root, x);
+
+        return this.root.getData().compareTo(x)==0 ? this.root:null;
+    }
+
+    private ArrayList<BinaryTree<T>> split(T x) {
+        BinaryTree<T> right;
+        BinaryTree<T> left;
+
+        if (x.compareTo(root.getData()) > 0) {
+            right = root.getRight();
+            left = root;
+            left.setRight(null);
+        } else {
+            left = root.getLeft();
+            right = root;
+            right.setLeft(null);
+        }
+        ArrayList<BinaryTree<T>> l_r = new ArrayList<>();
+        l_r.add(left);
+        l_r.add(right);
+
+        return l_r;
+    }
+
+    public BinaryTree<T> insert(T x) {
+        if (root == null) {
+            root = new BinaryTree<>(x);
+            return root;
+        }
+        splay(x);
+
+        ArrayList<BinaryTree<T>> l_r = split(x);
+
+        BinaryTree<T> left = l_r.get(0);
+        BinaryTree<T> right = l_r.get(1);
+
+        root = new BinaryTree<>(x);
+        root.setLeft(left);
+        root.setRight(right);
+
+
+        return root;
+    }
+
+    public BinaryTree<T> delete(T x) {
+        if (root == null) return null;
+
+        BinaryTree<T> searchResult = splay(x);
+
+        if (searchResult.getData().compareTo(x) != 0)
+            return null;
+
+        BinaryTree<T> leftSubtree = root.getLeft();
+        BinaryTree<T> rightSubtree = root.getRight();
+
+        // Set the 'to be deleted' key ready for garbage collection
+        root.setLeft(null);
+        root.setRight(null);
+        root.setData(null);
+
+        root = join(leftSubtree, rightSubtree);
+
+        return root;
+    }
+
+    private BinaryTree<T> join(BinaryTree<T> L, BinaryTree<T> R) {
+
+        if (L == null) {
+            root = R;
+            return R;
+        }
+        root = splayUtil(L, findMax(L));
+        root.setRight(R);
+        return root;
+    }
+    public T findMax(BinaryTree<T> root){
```

Comment:
Can you also add findMin which is exactly the same but digs left into the tree.

Reply:
Oke Sure